man/boobcoming.1
.\" -*- coding: utf-8 -*-
.\" This file was generated automatically by tools/make_man.sh.
.TH BOOBCOMING 1 "13 February 2020" "boobcoming 2\&.0"
.SH NAME
boobcoming \- see upcoming events
.SH SYNOPSIS
.B boobcoming
[\-dqv] [\-b \fIbackends\fR] [\-cnfs] [\fIcommand\fR [\fIarguments\fR..]]
.br
.B boobcoming
[\-\-help] [\-\-version]

.SH DESCRIPTION
.LP

Console application to see upcoming events.

.SS Supported websites:
* agendaculturel (agendaculturel website)
.br
* agendadulibre (agendadulibre website)
.br
* allocine (AlloCiné French cinema database service)
.br
* billetreduc (BilletReduc discount ticket reservation)
.br
* centquatre (centquatre website)
.br
* hybride (hybride website)
.br
* lucca (Lucca RH)
.br
* pariskiwi (ParisKiwi website)
.br
* razibus (site annonçant les évènements attendus par les punks a chiens)
.br
* residentadvisor (residentadvisor website)
.br
* senscritique (senscritique website)
.br
* sueurdemetal (SueurDeMetal French concerts list website)
.SH BOOBCOMING COMMANDS
.TP
\fBattends\fR \fIID\fR1 [\fIID\fR2 \fIID\fR3 ...]
.br
Register as participant of an event.
.br
ID is the identifier of the event.
.TP
\fBdebug\fR
.br
Launch a debug Python shell
.TP
\fBexport\fR \fIFILENAME\fR [\fIID\fR1 \fIID\fR2 \fIID\fR3 ...]
.br
ID is the identifier of the event. If no ID every events are exported
.br

.br
FILENAME is where to write the file. If FILENAME is '\-', the file is written to stdout.
.br

.br
Export event in ICALENDAR format
.TP
\fBinfo\fR \fIID\fR
.br
Get information about an event.
.TP
\fBlist\fR [\fIPATTERN\fR]List upcoming events, pattern can be an english or french week day, 'today' or a date (dd/mm/yy[yy])
.br

.br
Default is limited to 10 results.
.TP
\fBload\fR [query name]without query name : list loadable queries
.br
with query name laod query
.br

.br
Default is limited to 10 results.
.TP
\fBsearch\fR
.br
search for an event. Parameters interactively asked
.br

.br
Default is limited to 10 results.
.TP
\fBunattends\fR \fIID\fR1 [\fIID\fR2 \fIID\fR3 ...]
.br
Unregister you participation for an event.
.br
ID is the identifier of the event.
.SH WEBOOB COMMANDS
.TP
\fBbackends\fR [\fIACTION\fR] [\fIBACKEND_NAME\fR]...
.br
Select used backends.
.br

.br
ACTION is one of the following (default: list):
.br
* enable         enable given backends
.br
* disable        disable given backends
.br
* only           enable given backends and disable the others
.br
* list           list backends
.br
* add            add a backend
.br
* register       register a new account on a website
.br
* edit           edit a backend
.br
* remove         remove a backend
.br
* list\-modules   list modules
.TP
\fBcd\fR [\fIPATH\fR]
.br
Follow a path.
.br
".." is a special case and goes up one directory.
.br
"" is a special case and goes home.
.TP
\fBcondition\fR [\fIEXPRESSION\fR | off]
.br
If an argument is given, set the condition expression used to filter the results. See CONDITION section for more details and the expression.
.br
If the "off" value is given, conditional filtering is disabled.
.br

.br
If no argument is given, print the current condition expression.
.TP
\fBcount\fR [\fINUMBER\fR | off]
.br
If an argument is given, set the maximum number of results fetched.
.br
NUMBER must be at least 1.
.br
"off" value disables counting, and allows infinite searches.
.br

.br
If no argument is given, print the current count value.
.TP
\fBformatter\fR [list | \fIFORMATTER\fR [\fICOMMAND\fR] | option \fIOPTION_NAME\fR [on | off]]
.br
If a FORMATTER is given, set the formatter to use.
.br
You can add a COMMAND to apply the formatter change only to
.br
a given command.
.br

.br
If the argument is "list", print the available formatters.
.br

.br
If the argument is "option", set the formatter options.
.br
Valid options are: header, keys.
.br
If on/off value is given, set the value of the option.
.br
If not, print the current value for the option.
.br

.br
If no argument is given, print the current formatter.
.TP
\fBlogging\fR [\fILEVEL\fR]
.br
Set logging level.
.br

.br
Availables: debug, info, warning, error.
.br
* quiet is an alias for error
.br
* default is an alias for warning
.TP
\fBls\fR [\-d] [\-\fIU\fR] [\fIPATH\fR]
.br
List objects in current path.
.br
If an argument is given, list the specified path.
.br
Use \-U option to not sort results. It allows you to use a "fast path" to
.br
return results as soon as possible.
.br
Use \-d option to display information about a collection (and to not
.br
display the content of it). It has the same behavior than the well
.br
known UNIX "ls" command.
.br

.br
Default is limited to 40 results.
.TP
\fBquit\fR
.br
Quit the application.
.TP
\fBselect\fR [\fIFIELD_NAME\fR]... | "$direct" | "$full"
.br
If an argument is given, set the selected fields.
.br
$direct selects all fields loaded in one http request.
.br
$full selects all fields using as much http requests as necessary.
.br

.br
If no argument is given, print the currently selected fields.

.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-b BACKENDS\fR, \fB\-\-backends=BACKENDS\fR
what backend(s) to enable (comma separated)
.TP
\fB\-e EXCLUDE_BACKENDS\fR, \fB\-\-exclude\-backends=EXCLUDE_BACKENDS\fR
what backend(s) to exclude (comma separated)
.TP
\fB\-I\fR, \fB\-\-insecure\fR
do not validate SSL
.TP
\fB\-\-nss\fR
Use NSS instead of OpenSSL
.TP
\fB\-\-auto\-update\fR
Automatically check for updates when a bug in a module is encountered

.SH LOGGING OPTIONS
.TP
\fB\-d\fR, \fB\-\-debug\fR
display debug messages. Set up it twice to more verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR
display only error messages
.TP
\fB\-v\fR, \fB\-\-verbose\fR
display info messages
.TP
\fB\-\-logging\-file=LOGGING_FILE\fR
file to save logs
.TP
\fB\-a\fR, \fB\-\-save\-responses\fR
save every response
.TP
\fB\-\-export\-session\fR
log browser session cookies after login

.SH RESULTS OPTIONS
.TP
\fB\-c CONDITION\fR, \fB\-\-condition=CONDITION\fR
filter result items to display given a boolean expression. See CONDITION section
for the syntax
.TP
\fB\-n COUNT\fR, \fB\-\-count=COUNT\fR
limit number of results (from each backends)
.TP
\fB\-s SELECT\fR, \fB\-\-select=SELECT\fR
select result item keys to display (comma separated)

.SH FORMATTING OPTIONS
.TP
\fB\-f FORMATTER\fR, \fB\-\-formatter=FORMATTER\fR
select output formatter (csv, htmltable, ical_formatter, json, json_line,
multiline, simple, simple_upcoming, table, upcoming, upcoming_list, webkit)
.TP
\fB\-\-no\-header\fR
do not display header
.TP
\fB\-\-no\-keys\fR
do not display item keys
.TP
\fB\-O OUTFILE\fR, \fB\-\-outfile=OUTFILE\fR
file to export result

.SH CONDITION
The \-c and \-\-condition is a flexible way to filter and get only interesting results. It supports conditions on numerical values, dates, durations and strings. Dates are given in YYYY\-MM\-DD or YYYY\-MM\-DD HH:MM format. Durations look like XhYmZs where X, Y and Z are integers. Any of them may be omitted. For instance, YmZs, XhZs or Ym are accepted.
The syntax of one expression is "\fBfield operator value\fR". The field to test is always the left member of the expression.
.LP
The field is a member of the objects returned by the command. For example, a bank account has "balance", "coming" or "label" fields.
.SS The following operators are supported:
.TP
=
Test if object.field is equal to the value.
.TP
!=
Test if object.field is not equal to the value.
.TP
>
Test if object.field is greater than the value. If object.field is date, return true if value is before that object.field.
.TP
<
Test if object.field is less than the value. If object.field is date, return true if value is after that object.field.
.TP
|
This operator is available only for string fields. It works like the Unix standard \fBgrep\fR command, and returns True if the pattern specified in the value is in object.field.
.SS Expression combination
.LP
You can make a expression combinations with the keywords \fB" AND "\fR, \fB" OR "\fR an \fB" LIMIT "\fR.
.LP
The \fBLIMIT\fR keyword can be used to limit the number of items upon which running the expression. \fBLIMIT\fR can only be placed at the end of the expression followed by the number of elements you want.
.SS Examples:
.nf
.B boobank ls \-\-condition 'label=Livret A'
.fi
Display only the "Livret A" account.
.PP
.nf
.B boobank ls \-\-condition 'balance>10000'
.fi
Display accounts with a lot of money.
.PP
.nf
.B boobank history account@backend \-\-condition 'label|rewe'
.fi
Get transactions containing "rewe".
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01 AND date<2013\-12\-09'
.fi
Get transactions betweens the 2th December and 8th December 2013.
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01  LIMIT 10'
.fi
Get transactions after the 2th December in the last 10 transactions

.SH COPYRIGHT
Copyright(C) 2012-2020 Bezleputh
.LP
For full copyright information see the COPYING file in the weboob package.
.LP
.RE
.SH FILES
"~/.config/weboob/backends" 

.SH SEE ALSO
Home page: http://weboob.org/applications/boobcoming


ChangeLog
Weboob 2.0 (2020-02-13)

	Applications: boobank
	* keep Value for TransferStep and AddRecipientStep
	* declare utf-8 in OFX formatter
	* give account object to iter_transfer_recipient

	Applications: boobill
	* using a persistent storage for browser state

	Applications: boobsize
	* fix name

	Applications: translaboob
	* use babel to detect language names and pass ISO-639 codes

	Applications: weboobdebug
	* weboob-debug: fix name

	Applications: weboobmain
	* move scripts/weboob to a dedicated dir in weboob/applications

	Applications: weboobrepos
	* weboob-repos: port to python3

	Browser
	* set a parent logger to AbstractElement
	* Allows SeleniumBrowser to use MANUAL proxies
	* AUTHENTICATION_METHODS set in init
	* delete prefix 'handle_' on authentication methods
	* clean authentication config keys after login
	* make variable name more understandable
	* make get_expire stronger
	* TWOFA_DURATION for the state expire date
	* better names
	* set expire from logged_date
	* make authentication methods more flexible
	* handle both 2FA and regular login methods
	* add possibility to clear specific cookies
	* handle NeedInteractiveFor2FA
	* add documentation about otp, sms and resume
	* add TwoFactorBrowser to handle multiple 2FA
	* Set default logging level to ERROR for selenium
	* Added the ability to overrride *Elements' env copy
	* Allow passing custom headers in Form.submit()
	* set attribute 'weboob' on browser
	* Normalize encodings
	* BrowserTooManyRequests: Remove class attributes for date helpers
	* BrowserTooManyRequests: Raise a TypeError if the next_try is not a datetime
	* BrowserTooManyRequests: Check whether the next_try is None after the date check
	* BrowserTooManyRequests: Add message parameter
	* [weboob/browser/browser+pages] Fixed doctest BASEURL
	* add some helpers to TooManyRequests
	* add rate limiting exception
	* add build_refresh_token_parameters method
	* python3 compatibility
	* send redirect_uri when using the refresh_token
	* allow multiple inheritance with AbstractBrowser
	* Implemented wait_xpath_invisible
	* Avoid crash when no BASEURL is set
	* add klass attribute to decorated @method
	* give state to the module at initialization
	* introduce BrowserParamURL for passing params implictly
	* Implement the empty_xpath attribute
	* accept new parameter proxy_headers (but don't use it)
	* Browser takes a new parameter 'proxy_headers'
	* if use of refresh_token fails, forget it
	* support when token_type key is missing
	* if access_token_expire is None, the access_token has no expiration
	* in case we got a 401 response, forget the access_token
	* fix compatibility with python3
	* Handle generic error for bad url callback
	* Repair refresh token
	* StatesMixin cookies should be str
	* OAuth2PKCEMixin codes shouldn't be bytes
	* Rewrite/improve PagesBrowser example

	Browser: Filters
	* Add the format .Italian() to CleanDecimal
	* New keyword argument to remove accented characters on CleanText filter.
	* Changed Capitalize filter to Title
	* Updated & Reordered __all__
	* Add dates to FormValue
	* Add color to FormValue
	* Add number/range type to FormValue, and tests
	* MultiJoin accepts types other than str
	* Regexp can use regex library
	* Allow all dateutil.parser arguments in DateTime
	* Strict dates by default
	* Catch incomplete dates
	* Explicit error message
	* Add MultiJoin
	* Join can use default argument
	* Add doctests to Join
	* TableCell(colspan=True) is now the only behavior desired
	* add FromTimestamp to convert to datetime
	* Fix error message for FormValue
	* Add new types for FormValue
	* Add MapIn filter, that allow the pattern of a selected value to another value using a dict.
	* Reorganize imports / add missing to __all__
	* Add Coalesce

	Capabilities
	* rename accepted_doc_types attribute of CapDocument
	* Add KIID to DocumentTypes
	* Added new capability CapBands
	* Keep only PER
	* add PER account type
	* HousingPhoto inherits BaseImage
	* use PostalAddress and GeoCoords types where possible
	* introduce weboob.capabilities.address for a few object types
	* Add firstname and lastname field
	* add capability_to_string function to get capability name
	* Add new SAR currency
	* fix missing TransferInvalidOTP export
	* add TransferInvalidOTP exception
	* use StringField standard default value for transfer beneficiary label and number
	* Indentation cosmetic changes
	* Add new Investment attributes
	* Add new document type 'income_tax'
	* Add field 'original_gross_amount' to Transaction
	* Add field 'gross_amount' to Transaction
	* Remove performance_history default value
	* Added new attributes for employee savings accounts
	* Add RecaptchaV3Job to captcha capability
	* add support for commission in currency in bank/Transaction
	* add new add recipient exception for invalid otp
	* add has_file field to Document class
	* Add the relationship between the credentials owner (PSU) and the account
	* Add bdate. Represent the date that the transaction appear in the website.
	* prevent casting id from None to 'None' in constructor

	Documentation
	* add pip install
	* fix compatibility with sphinx 2.0
	* do not use deprecated defindex.html anymore
	* Use correct word

	General
	* New airparif module (CapGauge)
	* New aviva module (CapBank, CapBankWealth)
	* New cesu module (CapDocument)
	* New degiro module (CapBank, CapBankWealth)
	* New hsbchk module (CapBank)
	* New lendosphere module (CapBank, CapBankWealth)
	* New lucca module (CapCalendarEvent, CapDocument)
	* New lunchr module (CapBank)
	* New metalarchives module (CapBands)
	* New pajemploi module (CapDocument)
	* New ticketscesu module (CapBank)
	* New wiseed module (CapBank, CapBankWealth)
	* Deleted biplan module
	* Deleted figgo module
	* Deleted lolix module
	* Deleted wellsfargo module

	General: Core
	* send request_information to modules in all case, except if stdout is not a tty
	* handle error NeedInteractiveFor2FA

weboob/applications/boobcoming/boobcoming.py
# -*- coding: utf-8 -*-

# Copyright(C) 2013 Bezleputh
#
# This file is part of weboob.
#
# weboob is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# weboob is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with weboob. If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

from datetime import time, datetime
from dateutil import tz

from weboob.tools.date import parse_date
from weboob.tools.application.formatters.iformatter import IFormatter, PrettyFormatter
from weboob.capabilities.base import empty
from weboob.capabilities.calendar import CapCalendarEvent, Query, CATEGORIES, BaseCalendarEvent, TICKET, STATUS
from weboob.tools.application.repl import ReplApplication, defaultcount
from weboob.tools.config.yamlconfig import YamlConfig

__all__ = ['Boobcoming']


class UpcomingSimpleFormatter(IFormatter):
    MANDATORY_FIELDS = ('id', 'start_date', 'category', 'summary', 'status')

    def format_obj(self, obj, alias):
        result = u'%s - %s' % (obj.backend, obj.category)
        if not empty(obj.start_date):
            result += u' - %s' % obj.start_date.strftime('%H:%M')
        result += u' - %s' % obj.summary
        if obj.status == STATUS.CANCELLED:
            result += u' (cancelled)'
        return result


class ICalFormatter(IFormatter):
    MANDATORY_FIELDS = ('id', 'start_date', 'end_date', 'summary', 'status')

    def start_format(self, **kwargs):
        result = u'BEGIN:VCALENDAR\r\n'
        result += u'VERSION:2.0\r\n'
        result += u'PRODID:-//hacksw/handcal//NONSGML v1.0//EN\r\n'
        self.output(result)

    def format_obj(self, obj, alias):
        result = u'BEGIN:VEVENT\r\n'

        utc_zone = tz.gettz('UTC')

        event_timezone = tz.gettz(obj.timezone)
        start_date = obj.start_date if not empty(obj.start_date) else datetime.now()
        if isinstance(start_date, datetime):
            start_date = start_date.replace(tzinfo=event_timezone)
            utc_start_date = start_date.astimezone(utc_zone)
            result += u'DTSTART:%s\r\n' % utc_start_date.strftime("%Y%m%dT%H%M%SZ")
        else:
            result += u'DTSTART:%s\r\n' % start_date.strftime("%Y%m%d")

        end_date = obj.end_date if not empty(obj.end_date) else datetime.combine(start_date, time.max)
        if isinstance(end_date, datetime):
            end_date = end_date.replace(tzinfo=event_timezone)
            utc_end_date = end_date.astimezone(utc_zone)
            result += u'DTEND:%s\r\n' % utc_end_date.strftime("%Y%m%dT%H%M%SZ")
        else:
            result += u'DTEND:%s\r\n' % end_date.strftime("%Y%m%d")

        result += u'SUMMARY:%s\r\n' % obj.summary
        result += u'UID:%s\r\n' % obj.id
        result += u'STATUS:%s\r\n' % obj.status

        location = ''
        if hasattr(obj, 'location') and not empty(obj.location):
            location += obj.location + ' '

        if hasattr(obj, 'city') and not empty(obj.city):
            location += obj.city + ' '

        if not empty(location):
            result += u'LOCATION:%s\r\n' % location

        if hasattr(obj, 'categories') and not empty(obj.categories):
            result += u'CATEGORIES:%s\r\n' % obj.categories

        if hasattr(obj, 'description') and not empty(obj.description):
            result += u'DESCRIPTION:%s\r\n' % obj.description.strip(' \t\n\r')\
                                                             .replace('\r', '')\
                                                             .replace('\n', r'\n')\
                                                             .replace(',', '\,')

        if hasattr(obj, 'transp') and not empty(obj.transp):
            result += u'TRANSP:%s\r\n' % obj.transp

        if hasattr(obj, 'sequence') and not empty(obj.sequence):
            result += u'SEQUENCE:%s\r\n' % obj.sequence

        if hasattr(obj, 'url') and not empty(obj.url):
            result += u'URL:%s\r\n' % obj.url

        result += u'END:VEVENT\r\n'
        return result

    def flush(self, **kwargs):
        self.output(u'END:VCALENDAR')


class UpcomingListFormatter(PrettyFormatter):
    MANDATORY_FIELDS = ('id', 'start_date', 'end_date', 'summary', 'category', 'status')

    def get_title(self, obj):
        return ' %s - %s ' % (obj.category, obj.summary)

    def get_description(self, obj):
        result = u''
        if not empty(obj.start_date):
            result += u'\tDate: %s\n' % obj.start_date.strftime('%A %d %B %Y')
            result += u'\tHour: %s' % obj.start_date.strftime('%H:%M')
            if not empty(obj.end_date):
                result += ' - %s' % obj.end_date.strftime('%H:%M')
                days_diff = (obj.end_date - obj.start_date).days
                if days_diff >= 1:
                    result += ' (%i day(s) later)' % (days_diff)
            result += '\n'
        if obj.status == STATUS.CANCELLED:
            result += '\tStatus: Cancelled\n'
        return result.strip('\n\t')


class UpcomingFormatter(IFormatter):
    MANDATORY_FIELDS = ('id', 'start_date', 'end_date', 'summary', 'category', 'status')

    def format_obj(self, obj, alias):
        result = u'%s%s - %s%s\n' % (self.BOLD, obj.category, obj.summary, self.NC)

        if not empty(obj.start_date):
            if not empty(obj.end_date):
                days_diff = (obj.end_date - obj.start_date).days
                if days_diff >= 1:
                    result += u'From: %s to %s ' % (obj.start_date.strftime('%A %d %B %Y'),
                                                    obj.end_date.strftime('%A %d %B %Y'))
                else:
                    result += u'Date: %s\n' % obj.start_date.strftime('%A %d %B %Y')
                    result += u'Hour: %s' % obj.start_date.strftime('%H:%M')
                    result += ' - %s' % obj.end_date.strftime('%H:%M')
            else:
                result += u'Date: %s\n' % obj.start_date.strftime('%A %d %B %Y')
                result += u'Hour: %s' % obj.start_date.strftime('%H:%M')

            result += '\n'

        if hasattr(obj, 'location') and not empty(obj.location):
            result += u'Location: %s\n' % obj.location

        if hasattr(obj, 'city') and not empty(obj.city):
            result += u'City: %s\n' % obj.city

        if hasattr(obj, 'event_planner') and not empty(obj.event_planner):
            result += u'Event planner: %s\n' % obj.event_planner

        if hasattr(obj, 'booked_entries') and not empty(obj.booked_entries) and \
           hasattr(obj, 'max_entries') and not empty(obj.max_entries):
            result += u'Entry: %s/%s \n' % (obj.booked_entries, obj.max_entries)
        elif hasattr(obj, 'booked_entries') and not empty(obj.booked_entries):
            result += u'Entry: %s \n' % (obj.booked_entries)
        elif hasattr(obj, 'max_entries') and not empty(obj.max_entries):
            result += u'Max entries: %s \n' % (obj.max_entries)

        if hasattr(obj, 'description') and not empty(obj.description):
            result += u'Description:\n %s\n\n' % obj.description

        if hasattr(obj, 'price') and not empty(obj.price):
            result += u'Price: %.2f\n' % obj.price

        if hasattr(obj, 'ticket') and not empty(obj.ticket):
            result += u'Ticket: %s\n' % obj.ticket

        if hasattr(obj, 'url') and not empty(obj.url):
            result += u'URL: %s\n' % obj.url

        if hasattr(obj, 'status') and not empty(obj.status):
            result += u'Status: %s\n' % obj.status

        return result


class Boobcoming(ReplApplication):
    APPNAME = 'boobcoming'
    VERSION = '2.1'
    COPYRIGHT = 'Copyright(C) 2012-YEAR Bezleputh'
    DESCRIPTION = "Console application to see upcoming events."
    SHORT_DESCRIPTION = "see upcoming events"
    CAPS = CapCalendarEvent
    EXTRA_FORMATTERS = {'upcoming_list': UpcomingListFormatter,
                        'upcoming': UpcomingFormatter,
                        'simple_upcoming': UpcomingSimpleFormatter,
                        'ical_formatter': ICalFormatter,
                        }
    COMMANDS_FORMATTERS = {'list': 'upcoming_list',
                           'search': 'upcoming_list',
                           'load': 'upcoming_list',
                           'ls': 'upcoming_list',
                           'info': 'upcoming',
                           'export': 'ical_formatter'
                           }

    def main(self, argv):
        self.load_config(klass=YamlConfig)
        return ReplApplication.main(self, argv)

    def comp_object(self, obj1, obj2):
        if isinstance(obj1, BaseCalendarEvent) and isinstance(obj2, BaseCalendarEvent):
            if obj1.start_date == obj2.start_date:
                return 0
            if obj1.start_date > obj2.start_date:
                return 1
            return -1
        else:
            return super(Boobcoming, self).comp_object(obj1, obj2)

    def select_values(self, values_from, values_to, query_str):
        r = 'notempty'
        while r != '':
            for i, value in enumerate(values_from, 1):
                print('  %s%2d)%s [%s] %s' % (self.BOLD,
                                              i,
                                              self.NC,
                                              'x' if value in values_to else ' ',
                                              value))
            r = self.ask(query_str, regexp='(\d+|)', default='')

            if not r.isdigit():
                continue
            r = int(r)
            if r <= 0 or r > len(values_from):
                continue
            value = list(values_from)[r - 1]
            if value in values_to:
                values_to.remove(value)
            else:
                values_to.append(value)

    @defaultcount(10)
    def do_search(self, line):
        """
        search

        search for an event. Parameters interactively asked
        """
        query = Query()
        self.select_values(CATEGORIES, query.categories, '  Select category (or empty to stop)')
        self.select_values(TICKET, query.ticket, '  Select tickets status (or empty to stop)')

        if query.categories and len(query.categories) > 0 and query.ticket and len(query.ticket) > 0:
            query.city = self.ask('Enter a city', default='')
            query.summary = self.ask('Enter a title', default='')

            start_date = self.ask_date('Enter a start date', default='today')
            end_date = self.ask_date('Enter a end date', default='')

            if end_date:
                if end_date == start_date:
                    end_date = datetime.combine(start_date, time.max)
                else:
                    end_date = datetime.combine(end_date, time.max)

            query.start_date = datetime.combine(start_date, time.min)
            query.end_date = end_date

            save_query = self.ask('Save query (y/n)', default='n')
            if save_query.upper() == 'Y':
                name = ''
                while not name:
                    name = self.ask('Query name')

                self.config.set('queries', name, query)
                self.config.save()
            self.complete_search(query)

    def complete_search(self, query):
        self.change_path([u'events'])
        self.start_format()
        for event in self.do('search_events', query):
            if event:
                self.cached_format(event)

    def ask_date(self, txt, default=''):
        r = self.ask(txt, default=default)
        return parse_date(r)

    @defaultcount(10)
    def do_load(self, query_name):
        """
        load [query name]
        without query name : list loadable queries
        with query name laod query
        """
        queries = self.config.get('queries')
        if not queries:
            print('There is no saved queries', file=self.stderr)
            return 2

        if not query_name:
            for name in queries.keys():
                print('  %s* %s %s' % (self.BOLD,
                                       self.NC,
                                       name))
            query_name = self.ask('Which one')

        if query_name in queries:
            self.complete_search(queries.get(query_name))
        else:
            print('Unknown query', file=self.stderr)
            return 2

    @defaultcount(10)
    def do_list(self, line):
        """
        list [PATTERN]
        List upcoming events, pattern can be an english or french week day, 'today' or a date (dd/mm/yy[yy])
        """

        self.change_path([u'events'])
        if line:
            _date = parse_date(line)
            if not _date:
                print('Invalid argument: %s' % self.get_command_help('list'), file=self.stderr)
                return 2

            date_from = datetime.combine(_date, time.min)
            date_to = datetime.combine(_date, time.max)
        else:
            date_from = datetime.now()
            date_to = None

        for event in self.do('list_events', date_from, date_to):
            self.cached_format(event)

    def complete_info(self, text, line, *ignored):
        args = line.split(' ')
        if len(args) == 2:
            return self._complete_object()

    def do_info(self, _id):
        """
        info ID

        Get information about an event.
        """

        if not _id:
            print('This command takes an argument: %s' % self.get_command_help('info', short=True), file=self.stderr)
            return 2

        event = self.get_object(_id, 'get_event')

        if not event:
            print('Upcoming event not found: %s' % _id, file=self.stderr)
            return 3

        self.start_format()
        self.format(event)

    def do_export(self, line):
        """
        export FILENAME [ID1 ID2 ID3 ...]

        ID is the identifier of the event. If no ID every events are exported

        FILENAME is where to write the file. If FILENAME is '-', the file is written to stdout.

        Export event in ICALENDAR format
        """
        if not line:
            print('This command takes at leat one argument: %s' % self.get_command_help('export'), file=self.stderr)
            return 2

        _file, args = self.parse_command_args(line, 2, req_n=1)

        l = self.retrieve_events(args)
        if l == 3:
            return 3

        if not _file == "-":
            dest = self.check_file_ext(_file)
            self.formatter.outfile = dest

        self.formatter.start_format()
        for item in l:
            self.format(item)

    def retrieve_events(self, args):
        l = []

        if not args:
            _ids = []
            for event in self.do('list_events', datetime.now(), None):
                _ids.append(event.id)
        else:
            _ids = args.strip().split(' ')

        for _id in _ids:
            event = self.get_object(_id, 'get_event')

            if not event:
                print('Upcoming event not found: %s' % _id, file=self.stderr)
                return 3

            l.append(event)

        return l

    def check_file_ext(self, _file):
        splitted_file = _file.split('.')
        if splitted_file[-1] != 'ics':
            return "%s.ics" % _file
        else:
            return _file

    def do_attends(self, line):
        """
        attends ID1 [ID2 ID3 ...]

        Register as participant of an event.
        ID is the identifier of the event.
        """
        if not line:
            print('This command takes at leat one argument: %s' % self.get_command_help('attends'), file=self.stderr)
            return 2

        args = self.parse_command_args(line, 1, req_n=1)

        l = self.retrieve_events(args[0])
        for event in l:
            # we wait till the work be done, else the errors are not handled
            self.do('attends_event', event, True).wait()

    def do_unattends(self, line):
        """
        unattends ID1 [ID2 ID3 ...]

        Unregister you participation for an event.
        ID is the identifier of the event.
        """

        if not line:
            print('This command takes at leat one argument: %s' % self.get_command_help('unattends'), file=self.stderr)
            return 2

        args = self.parse_command_args(line, 1, req_n=1)

        l = self.retrieve_events(args[0])
        for event in l:
            self.do('attends_event', event, False)
